### PR TITLE
GH-43502: [Java] Fix Java JNI / AMD64 manylinux2014 Java JNI test not test dataset module

### PR DESCRIPTION
**Pull request by vibhatha**
### Rationale for this change

JNI tests had a typo in the script where we have included `gandiva` instead of `dataset`. 

### What changes are included in this PR?

This PR fixes that typo in the current version. 

### Are these changes tested?

Yes, by existing tests and CIs. 

### Are there any user-facing changes?

No
* GitHub Issue: apache/arrow#43502
* GitHub Issue: #43502

### CODE REVIEW

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

**Comment by jinchengchenghh** [comment]
And it still has a problem , `https://github.com/apache/arrow/blob/main/ci/docker/java-jni-manylinux-201x.dockerfile#L53`, `ARROW_JAVA_JNI` set, so jni build exists, but jni test detect ARROW_DATASET

**Comment by vibhatha** [comment]
> And it still has a problem , https://github.com/apache/arrow/blob/main/ci/docker/java-jni-manylinux-201x.dockerfile#L53, `ARROW_JAVA_JNI` set, so jni build exists, but jni test detect ARROW_DATASET

I didn't get it clearly, could you please elaborate a bit?

**Comment by jinchengchenghh** [comment]
It check the `ARROW_DATASET`, but it is `OFF` now, so as `ARROW_GANDIVA` https://github.com/apache/arrow/blob/main/ci/scripts/java_test.sh#L41

**Comment by vibhatha** [comment]
> It check the `ARROW_DATASET`, but it is `OFF` now, so as `ARROW_GANDIVA` https://github.com/apache/arrow/blob/main/ci/scripts/java_test.sh#L41

Let me check, though if this is true, we have had the CIs running without any of these checks :thinking:

**Comment by vibhatha** [comment]
@jinchengchenghh on top of that there is already a CI failure for dataset.

**Comment by jinchengchenghh** [comment]
> > It check the `ARROW_DATASET`, but it is `OFF` now, so as `ARROW_GANDIVA` https://github.com/apache/arrow/blob/main/ci/scripts/java_test.sh#L41
> 
> Let me check, though if this is true, we have had the CIs running without any of these checks 🤔

I think you are right, not sure where to set the environment, but the test really triggers.

I will try to fix the exception, my IDE sync cause some problems

**Comment by jinchengchenghh** [comment]
Can you mark it as ignore?  I will draft a new PR to solve it.

**Comment by vibhatha** [comment]
> > > It check the `ARROW_DATASET`, but it is `OFF` now, so as `ARROW_GANDIVA` https://github.com/apache/arrow/blob/main/ci/scripts/java_test.sh#L41
> > 
> > 
> > Let me check, though if this is true, we have had the CIs running without any of these checks 🤔
> 
> I think you are right, not sure where to set the environment, but the test really triggers.

I will take a look at that. 

> I will try to fix the exception, my IDE sync cause some problems

Thanks, that would be wonderful, but I am not sure if this is related to the recent PRs (didn't check yet).

**Comment by vibhatha** [comment]
> Can you mark it as ignore? I will draft a new PR to solve it.

would you like to push the patch to this branch?

**Comment by jinchengchenghh** [comment]
https://github.com/apache/arrow/issues/43506

**Comment by vibhatha** [comment]
Btw, does that failing test in the CI work locally?

**Comment by jinchengchenghh** [comment]
Do you know how to push to this branch? I don't think I have write access to `vibhatha:gh-43502`

**Comment by jinchengchenghh** [comment]
https://github.com/apache/arrow/pull/41646 skip tests `java-jars`, maybe we still need environment `ARROW_DATASET` to trigger it in Java JNI / AMD64 manylinux2014 Java JNI test

**Comment by vibhatha** [comment]
> Do you know how to push to this branch? I don't think I have write access to `vibhatha:gh-43502`

I will add the rights.

**Comment by vibhatha** [comment]
@jinchengchenghh I have added you and you should have commit rights, please check.

**Comment by jinchengchenghh** [comment]
ok thanks! I will draft a PR as soon as possible

**Comment by jinchengchenghh** [comment]
I investigate the question again, the exception should not occur, looks like the linked library is not right.
```
TestFragmentScanOptions.testCsvConvertOptions:80 expected: <[Id: Int(32, true), Name: Utf8, Language: Utf8]> but was: <[Id;Name;Language: Utf8]>
```
Can anyone help to check it？

**Comment by vibhatha** [comment]
> I investigate the question again, the exception should not occur, looks like the linked library is not right.
> 
> ```
> TestFragmentScanOptions.testCsvConvertOptions:80 expected: <[Id: Int(32, true), Name: Utf8, Language: Utf8]> but was: <[Id;Name;Language: Utf8]>
> ```
> 
> Can anyone help to check it？

let me check this locally.

**Comment by vibhatha** [comment]
@github-actions crossbow submit java-jars

**Comment by vibhatha** [comment]
@jinchengchenghh has gandiva been set any other place?

**Comment by vibhatha** [comment]
@kou setting `ARROW_JAVA_JNI` should enable gandiva, dataset, etc, right?

File: ci/docker/java-jni-manylinux-201x.dockerfile [hunk 1]
```diff
@@ -52,4 +52,5 @@ RUN PYTHON=python /arrow/ci/scripts/install_gcs_testbench.sh default
 ENV ARROW_HOME=/tmp/local \
     ARROW_JAVA_CDATA=ON \
     ARROW_JAVA_JNI=ON \
-    ARROW_USE_CCACHE=ON
+    ARROW_USE_CCACHE=ON \
+    ARROW_DATASET=ON
```
Comment (kou) [inline ci/docker/java-jni-manylinux-201x.dockerfile:56]: Could you keep this list in alphabetical order?

**Review by kou — COMMENTED** [review]
> setting `ARROW_JAVA_JNI` should enable gandiva, dataset, etc, right?

Yes:

* https://github.com/apache/arrow/blob/0dec116d83e4160aa3387fa87be2a99ac0fd3390/ci/scripts/java_build.sh#L97-L99
* https://github.com/apache/arrow/blob/0dec116d83e4160aa3387fa87be2a99ac0fd3390/java/pom.xml#L877-L884

**Comment by vibhatha** [comment]
@jinchengchenghh we may not need to update the flags explicitly.

File: ci/docker/java-jni-manylinux-201x.dockerfile (hunk 1 above)
Comment (vibhatha) [reply ci/docker/java-jni-manylinux-201x.dockerfile:56]: @kou here we don't need the `ARROW_DATASET=ON`, right?

Comment (kou) [reply ci/docker/java-jni-manylinux-201x.dockerfile:56]: Right.

How about the following?

```diff
diff --git a/ci/scripts/java_test.sh b/ci/scripts/java_test.sh
index dd483ff254..5efda4318f 100755
--- a/ci/scripts/java_test.sh
+++ b/ci/scripts/java_test.sh
@@ -38,14 +38,10 @@ pushd ${source_dir}
 ${mvn} clean test
 
 projects=()
-if [ "${ARROW_DATASET}" = "ON" ]; then
-  projects+=(gandiva)
-fi
-if [ "${ARROW_GANDIVA}" = "ON" ]; then
-  projects+=(gandiva)
-fi
-if [ "${ARROW_ORC}" = "ON" ]; then
+if [ "${ARROW_JAVA_JNI}" = "ON" ]; then
   projects+=(adapter/orc)
+  projects+=(dataset)
+  projects+=(gandiva)
 fi
 if [ "${#projects[@]}" -gt 0 ]; then
   ${mvn} clean test \
```

Comment (vibhatha) [reply ci/docker/java-jni-manylinux-201x.dockerfile:56]: Perfect!!!

Comment (vibhatha) [reply ci/docker/java-jni-manylinux-201x.dockerfile:56]: I will add this. Thanks @kou

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

File: ci/docker/java-jni-manylinux-201x.dockerfile (hunk 1 above)
Comment (vibhatha) [reply ci/docker/java-jni-manylinux-201x.dockerfile]: @kou I added the suggestion and started the crossbows.

**Comment by jinchengchenghh** [comment]
I run the test locally, and failed, so it does not matter library.
I need some time to investigate it why the option `delimiter` set but make no effect.
I would suggest to mark the test as ignore to merge this PR first.

**Comment by vibhatha** [comment]
> I run the test locally, and failed, so it does not matter library. I need some time to investigate it why the option `delimiter` set but make no effect. I would suggest to mark the test as ignore to merge this PR first.

I would rather not ignore the test failures when merging the PR. Let's take our time and resolve it cleanly. If it requires making another PR, that's fine. But it would be wonderful if we can push it here.

**Comment by vibhatha** [comment]
@jinchengchenghh let me know your thoughts, and I really appreciate your push to fix this issue. If you need me to take a look, I can definitely help. :slightly_smiling_face:

**Comment by jinchengchenghh** [comment]
Much thanks for your help, I have found the error, I will push it to this branch later.

**Comment by vibhatha** [comment]
> Much thanks for your help, I have found the error, I will push it to this branch later.

Awesome!

File: java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java [hunk 2]
```diff
@@ -59,24 +71,29 @@ public void testCsvConvertOptions() throws Exception {
...
         CDataDictionaryProvider provider = new CDataDictionaryProvider()) {
       Data.exportSchema(allocator, schema, provider, cSchema);
-      CsvConvertOptions convertOptions = new CsvConvertOptions(ImmutableMap.of("delimiter", ";"));
-      convertOptions.setArrowSchema(cSchema);
-      CsvFragmentScanOptions fragmentScanOptions =
-          new CsvFragmentScanOptions(convertOptions, ImmutableMap.of(), ImmutableMap.of());
+      Data.exportSchema(allocator, schema, provider, cSchema2);
+      CsvFragmentScanOptions fragmentScanOptions1 =
+          create(cSchema, ImmutableMap.of(), ImmutableMap.of(), ImmutableMap.of("delimiter", ";"));
+      CsvFragmentScanOptions fragmentScanOptions2 =
```
Comment (vibhatha) [inline java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:80]: Why would we use `fragmentScanOptions2`? Instead we should be able to use `fragmentScanOptions1` (possibly rename to `fragmentScanOptions`)?

File: java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java [hunk 3]
```diff
@@ -102,30 +119,38 @@ public void testCsvConvertOptionsDelimiterNotSet() throws Exception {
...
-
-        assertEquals(schema.getFields(), reader.getVectorSchemaRoot().getSchema().getFields());
         int rowCount = 0;
         while (reader.loadNextBatch()) {
-          final ValueIterableVector<Integer> idVector =
-              (ValueIterableVector<Integer>) reader.getVectorSchemaRoot().getVector("Id");
-          assertThat(idVector.getValueIterable(), IsIterableContainingInOrder.contains(1, 2, 3));
+          final ValueIterableVector<Text> idVector =
+              (ValueIterableVector<Text>)
+                  reader.getVectorSchemaRoot().getVector("Id;Name;Language");
```
Comment (vibhatha) [inline java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:149]: Is this change necessary?

Comment (vibhatha) [reply java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:149]: By the way, we need to figure out what caused this test failure in the first place. One thing is since the CIs didn't run as expected we wouldn't have detected this during previous PRs, but we could have counted towards at least working locally. If this is failing locally, we need to see why it is happening. I only see a change to test cases :thinking:

File: java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java (hunk 2 above)
Comment (jinchengchenghh) [reply java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:80]: Because ArrowSchema can only be used once, otherwise it will raise exception ArrowSchema is released

File: java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java (hunk 3 above)
Comment (jinchengchenghh) [reply java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:149]: My fault, I does not run the unit test in local, only test the code integrated by my project.
I believe the CI would run the test, but not.

**Comment by jinchengchenghh** [comment]
Dataset tests are fixed, but there is gandiva test failed by
```
Error:  org.apache.arrow.gandiva.evaluator.ProjectorTest.testIsNull -- Time elapsed: 0.026 s <<< ERROR!
Failed to make LLVM module due to Could not create LLJIT instance: Symbols not found: [ llvm_orc_registerEHFrameSectionWrapper ]
```

File: java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java (hunk 3 above)
Comment (vibhatha) [reply java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:149]: Oh, we should definitely test them locally before pushing to the branch. CIs will always catch this, but unfortunately not in our case. Let's fix this and move forward.

**Comment by vibhatha** [comment]
> Dataset tests are fixed, but there is gandiva test failed by
> 
> ```
> Error:  org.apache.arrow.gandiva.evaluator.ProjectorTest.testIsNull -- Time elapsed: 0.026 s <<< ERROR!
> Failed to make LLVM module due to Could not create LLJIT instance: Symbols not found: [ llvm_orc_registerEHFrameSectionWrapper ]
> ```

Yeah I saw that too. We can probably ignore it for this PR, but shall we file an issue for this?

**Comment by jinchengchenghh** [comment]
Yes, I think so. https://github.com/apache/arrow/pull/43503#issuecomment-2264523717

File: java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java (hunk 3 above)
Comment (jinchengchenghh) [reply java/dataset/src/test/java/org/apache/arrow/dataset/TestFragmentScanOptions.java:149]: Yes I think so.

**Comment by lidavidm** [comment]
@github-actions crossbow submit -g java

**Comment by kou** [comment]
Can we use LLVM 16 or earlier for now?
See also: #39695

**Comment by vibhatha** [comment]
@kou I am not very familiar with the LLVM usage within the project. Although, are you recommending to make this change in this PR itself?

**Comment by kou** [comment]
Could you share the URL of the error?

**Comment by vibhatha** [comment]
> Could you share the URL of the error?

The CI failure for Gandiva: https://github.com/apache/arrow/actions/runs/10209549505/job/28247705201?pr=43503#step:7:62723

**Comment by kou** [comment]
Thanks.

https://github.com/apache/arrow/actions/runs/10209549505/job/28247705201?pr=43503#step:7:23897

```text
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.222 s <<< FAILURE! -- in org.apache.arrow.gandiva.evaluator.FilterProjectTest
Error:  org.apache.arrow.gandiva.evaluator.FilterProjectTest.testSimpleSV16 -- Time elapsed: 1.212 s <<< ERROR!
Failed to make LLVM module due to Could not create LLJIT instance: Symbols not found: [ llvm_orc_registerEHFrameSectionWrapper ]

	at org.apache.arrow.gandiva.evaluator.JniWrapper.buildFilter(Native Method)
	at org.apache.arrow.gandiva.evaluator.Filter.make(Filter.java:121)
	at org.apache.arrow.gandiva.evaluator.Filter.make(Filter.java:65)
	at org.apache.arrow.gandiva.evaluator.FilterProjectTest.testSimpleSV16(FilterProjectTest.java:51)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

**Comment by kou** [comment]
LLVM 17 was used:

https://github.com/apache/arrow/actions/runs/10209549505/job/28247705201?pr=43503#step:7:902

```text
 -- Found LLVMAlt: 17.0.2
-- Using LLVMConfig.cmake in: /opt/vcpkg/installed/amd64-linux-static-release/share/llvm
-- Found llvm-link /opt/vcpkg/installed/amd64-linux-static-release/tools/llvm/llvm-link
-- Found clang /opt/vcpkg/installed/amd64-linux-static-release/tools/llvm/clang-17
```

**Comment by vibhatha** [comment]
> LLVM 17 was used:
> 
> https://github.com/apache/arrow/actions/runs/10209549505/job/28247705201?pr=43503#step:7:902
> 
> ```
>  -- Found LLVMAlt: 17.0.2
> -- Using LLVMConfig.cmake in: /opt/vcpkg/installed/amd64-linux-static-release/share/llvm
> -- Found llvm-link /opt/vcpkg/installed/amd64-linux-static-release/tools/llvm/llvm-link
> -- Found clang /opt/vcpkg/installed/amd64-linux-static-release/tools/llvm/clang-17
> ```

@kou Thanks for looking into this. Shall we pick 16 as you suggested?

**Comment by kou** [comment]
I tried the following but it seems that it doesn't work...

```diff
diff --git a/ci/vcpkg/vcpkg.json b/ci/vcpkg/vcpkg.json
index 58b1382d1c..9dea31ffa1 100644
--- a/ci/vcpkg/vcpkg.json
+++ b/ci/vcpkg/vcpkg.json
@@ -4,6 +4,7 @@
   "description": "Cross-language development platform for in-memory analytics",
   "homepage": "https://arrow.apache.org",
   "supports": "x64 | (arm64 & !windows)",
+  "builtin-baseline": "943c5ef1c8f6b5e6ced092b242c8299caae2ff01",
   "dependencies": [
     "boost-filesystem",
     {
@@ -143,5 +144,11 @@
         }
       ]
     }
-  }
+  },
+  "overrides": [
+    {
+      "name": "llvm",
+      "version": "15.0.7"
+    }
+  ]
 }
```

**Comment by kou** [comment]
We can work on it as a separated issue.

**Comment by kou** [comment]
We can skip the Gandiva test for now.
(But it should be fixed before 18.0.0.)

**Comment by vibhatha** [comment]
> We can skip the Gandiva test for now. (But it should be fixed before 18.0.0.)

You mean merge this ignoring Gandiva failure, but not to disable it right?

**Comment by kou** [comment]
> You mean merge this ignoring Gandiva failure, but not to disable it right?

No. We disable only the failed tests for now in this PR to keep CI green. But we should enable the disabled tests again and fix them before we release a new version.

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

**Review by lidavidm — APPROVED** [review]

**Comment by vibhatha** [comment]
@kou I updated as you suggested, also created an [issue](https://github.com/apache/arrow/issues/43576) to track and update once the LLVM issue is resolved.

**Comment by danepitkin** [comment]
@github-actions crossbow submit java-jars

**Comment by vibhatha** [comment]
@danepitkin shall we re-run the java-jars? I am not sure what that failure is.

**Comment by kou** [comment]
Hmm. self-hosted arm64 Linux runner may have a problem: https://github.com/apache/arrow/blob/2e1b2452e70145b7d3a7f85df49e4c3ae4d2633b/dev/tasks/java-jars/github.yml#L43
Disk size?

**Comment by vibhatha** [comment]
@kou honestly not sure...

**Comment by danepitkin** [comment]
I think the previous java-jars successful run proves its fine. However the Java JNI CI job is failing. Did we want to fix the TestCsvFragment options in this PR?

**Comment by vibhatha** [comment]
> I think the previous java-jars successful run proves its fine. However the Java JNI CI job is failing. Did we want to fix the TestCsvFragment options in this PR?

Wasn't the objective in this PR though.

**Comment by danepitkin** [comment]
@github-actions crossbow submit java-jars

**Review by danepitkin — APPROVED** [review]
Gandiva builds, java-jars now failing on https://github.com/apache/arrow/issues/43506

**Comment by vibhatha** [comment]
Thanks @danepitkin
